File: TimberCruisingApp/Common/ForestixLogger.swift
// Phase 7 — local-only analytics + structured logging.
//
// ## Privacy model (spec Phase 7): "Collect local crash log only (no server upload)"
//
// Every event goes to two sinks:
//   1. `os_log` (Console.app / Instruments) — structured category "Forestix"
//   2. a rotating JSONL file at
//      `Application Support/Forestix/logs/events.jsonl`
//
// Nothing is ever sent to the network. The JSONL file is bounded to 10 MB:
// when we would exceed, we rotate to `events.prev.jsonl` and start fresh.
// Settings exports the current log via the share sheet.
//
// ## What we log
// Only project / plot / tree UUIDs and numeric stats. Never:
//   • owner names
//   • free-text notes
//   • photo or scan file paths
//   • GPS coordinates below 3 decimal places (city-block resolution)

import Foundation
import os.log

public enum ForestixLogger {

    // MARK: - Public event surface

    public enum Event: Sendable {
        case plotOpened(plotId: UUID, projectId: UUID)
        case backupCreated(projectId: UUID, bytes: Int64)
        case backupRestored(projectId: UUID, fromPath: String)
        case crashRecoveryPrompted(projectId: UUID, plotId: UUID)
    }

    // MARK: - Logging

    public static func log(_ event: Event,
                           file: String = #file,
                           line: Int = #line) {
        let dict = payload(for: event)
        Sink.shared.write(dict)
    }

    // MARK: - Log access (Settings → Export)

    public static var currentLogURL: URL { Sink.shared.currentURL }
    public static var previousLogURL: URL { Sink.shared.previousURL }

    /// Atomically clears both current and rotated logs. Called from the
    /// Settings "Clear analytics" button.
    public static func clear() { Sink.shared.clear() }

    // MARK: - Private payload builder

    private static func payload(for event: Event) -> [String: Any] {
        let now = ISO8601DateFormatter().string(from: Date())
        var base: [String: Any] = ["t": now]
        switch event {
        case .plotOpened(let pid, let proj):
            base["event"] = "plot.opened"
            base["plotId"] = pid.uuidString
            base["projectId"] = proj.uuidString
        case .backupCreated(let pid, let bytes):
            base["event"] = "backup.created"
            base["projectId"] = pid.uuidString
            base["bytes"] = bytes
        case .backupRestored(let pid, let from):
            base["event"] = "backup.restored"
            base["projectId"] = pid.uuidString
            // `fromPath` gets its last path component only — never log
            // user home paths.
            base["fromName"] = (from as NSString).lastPathComponent
        case .crashRecoveryPrompted(let pid, let plot):
            base["event"] = "crash.recovery.prompted"
            base["projectId"] = pid.uuidString
            base["plotId"] = plot.uuidString
        }
        return base
    }
}

// MARK: - Sink (file rotation + os_log)

private final class Sink {

    static let shared = Sink()
    private let queue = DispatchQueue(label: "forestix.logger",
                                      qos: .utility)
    private let osLog = OSLog(subsystem: "com.forestix", category: "Forestix")
    private let rotationThreshold: Int64 = 10 * 1024 * 1024  // 10 MB

    let currentURL: URL
    let previousURL: URL

    init() {
        let fm = FileManager.default
        let base: URL
        if let appSup = try? fm.url(for: .applicationSupportDirectory,
                                    in: .userDomainMask,
                                    appropriateFor: nil, create: true) {
            base = appSup.appendingPathComponent("Forestix/logs",
                                                 isDirectory: true)
        } else {
            base = fm.temporaryDirectory.appendingPathComponent("Forestix/logs",
                                                                 isDirectory: true)
        }
        try? fm.createDirectory(at: base, withIntermediateDirectories: true)
        self.currentURL  = base.appendingPathComponent("events.jsonl")
        self.previousURL = base.appendingPathComponent("events.prev.jsonl")
    }

    func write(_ dict: [String: Any]) {
        queue.async { [currentURL, previousURL, rotationThreshold, osLog] in
            guard JSONSerialization.isValidJSONObject(dict),
                  let data = try? JSONSerialization.data(
                    withJSONObject: dict,
                    options: [.sortedKeys]) else { return }

            // Mirror to os_log (structured category, Console-friendly).
            if let json = String(data: data, encoding: .utf8) {
                os_log("%{public}@", log: osLog, type: .info, json)
            }

            let fm = FileManager.default
            if !fm.fileExists(atPath: currentURL.path) {
                fm.createFile(atPath: currentURL.path, contents: nil)
            }

            // Rotation.
            if let sz = (try? fm.attributesOfItem(atPath: currentURL.path))?[.size]
                as? NSNumber, sz.int64Value > rotationThreshold {
                try? fm.removeItem(at: previousURL)
                try? fm.moveItem(at: currentURL, to: previousURL)
                fm.createFile(atPath: currentURL.path, contents: nil)
            }

            if let handle = try? FileHandle(forWritingTo: currentURL) {
                defer { try? handle.close() }
                _ = try? handle.seekToEnd()
                var line = data
                line.append(0x0A)           // '\n'
                try? handle.write(contentsOf: line)
            }
        }
    }

    func clear() {
        queue.sync { [currentURL, previousURL] in
            let fm = FileManager.default
            try? fm.removeItem(at: currentURL)
            try? fm.removeItem(at: previousURL)
        }
    }
}

File: android/app/src/main/java/com/hcjeong/forestix/common/ForestixLogger.kt
// Local-only analytics + structured logging — Android counterpart of iOS
// Common/ForestixLogger.swift (Phase 7).
//
// ## Privacy model: "Collect local crash log only (no server upload)"
//
// Every event is appended as one JSON line to a rotating JSONL file at
//   <filesDir>/logs/events.jsonl
// Nothing is ever sent to the network. The file is bounded to 10 MB: when a
// write would exceed it, we rotate to `events.prev.jsonl` and start fresh.
// Settings can share the current log and clear both files.
//
// ## What we log
// Only project / plot UUIDs and numeric stats. Never owner names, free-text
// notes, photo/scan paths, or precise coordinates.
//
// ## Event surface (locked)
// EXACTLY four events, matching the iOS sibling's kept set:
//   • plotOpened            — a plot is opened (created or resumed)
//   • backupCreated         — a full project export/backup artefact is written
//   • backupRestored        — a project backup is restored (no Android UI yet)
//   • crashRecoveryPrompted — the crash-recovery resume prompt is shown
//
// iOS's logger is a stateless static enum. The Android file writer needs an
// app Context, so `init(context)` is called once from AppEnvironment; after
// that the emit methods are Context-free, just like the iOS call sites. Writes
// run on a single background thread so the field UI never blocks on IO
// (mirroring the iOS serial `forestix.logger` queue).

package com.hcjeong.forestix.common

import android.content.Context
import android.net.Uri
import android.util.Log
import androidx.core.content.FileProvider
import org.json.JSONObject
import java.io.File
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone
import java.util.UUID
import java.util.concurrent.Executors

object ForestixLogger {

    private const val TAG = "Forestix"
    private const val ROTATION_THRESHOLD_BYTES = 10L * 1024 * 1024  // 10 MB

    @Volatile private var appContext: Context? = null

    private val writeExecutor = Executors.newSingleThreadExecutor { r ->
        Thread(r, "forestix.logger").apply { isDaemon = true }
    }
    private val iso8601 = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'", Locale.US).apply {
        timeZone = TimeZone.getTimeZone("UTC")
    }

    /// Bind the application context once (from AppEnvironment). Idempotent.
    fun init(context: Context) {
        if (appContext == null) appContext = context.applicationContext
    }

    // MARK: - Public event surface (exactly four events)

    /// A plot was opened — created fresh or resumed from the crash-recovery
    /// prompt. iOS event "plot.opened".
    fun plotOpened(plotId: UUID, projectId: UUID) {
        log {
            put("event", "plot.opened")
            put("plotId", plotId.toString())
            put("projectId", projectId.toString())
        }
    }

    /// A full project backup/export artefact was written to disk. iOS event
    /// "backup.created".
    fun backupCreated(projectId: UUID, bytes: Long) {
        log {
            put("event", "backup.created")
            put("projectId", projectId.toString())
            put("bytes", bytes)
        }
    }

    /// A project backup was restored. iOS event "backup.restored" — logs the
    /// source file's last path component only, never a full user path.
    fun backupRestored(projectId: UUID, fromPath: String) {
        log {
            put("event", "backup.restored")
            put("projectId", projectId.toString())
            put("fromName", File(fromPath).name)
        }
    }

    /// The crash-recovery resume prompt was shown for a candidate open plot.
    /// iOS event "crash.recovery.prompted".
    fun crashRecoveryPrompted(projectId: UUID, plotId: UUID) {
        log {
            put("event", "crash.recovery.prompted")
            put("projectId", projectId.toString())
            put("plotId", plotId.toString())
        }
    }

    // MARK: - Log access (Settings → share / clear)

    /// Current events log file, or null before init(). May not exist yet if
    /// nothing has been logged.
    fun currentLogFile(): File? = logsDir()?.let { File(it, "events.jsonl") }

    /// Rotated (previous) events log file, or null before init().
    fun previousLogFile(): File? = logsDir()?.let { File(it, "events.prev.jsonl") }

    /// True when there is a non-empty events log to share/clear.
    fun hasEvents(): Boolean {
        val f = currentLogFile() ?: return false
        return f.exists() && f.length() > 0L
    }

    /// Copy the events log into the shared-export cache and return a shareable
    /// content Uri (same FileProvider flow as ResearchExport.write — the only
    /// provider-configured path is `cacheDir/Exports/`).
    fun exportUri(context: Context): Uri? {
        val src = currentLogFile() ?: return null
        if (!src.exists() || src.length() == 0L) return null
        val dir = File(context.cacheDir, "Exports").apply { mkdirs() }
        val dst = File(dir, "forestix_events.jsonl")
        src.copyTo(dst, overwrite = true)
        return FileProvider.getUriForFile(context, "${context.packageName}.fileprovider", dst)
    }

    /// Clears both current and rotated logs.
    fun clear() {
        writeExecutor.execute {
            currentLogFile()?.delete()
            previousLogFile()?.delete()
        }
    }

    // MARK: - Private

    private fun logsDir(): File? =
        appContext?.let { File(it.filesDir, "logs").apply { mkdirs() } }

    private inline fun log(build: JSONObject.() -> Unit) {
        if (appContext == null) return
        val obj = JSONObject().apply {
            put("t", iso8601.format(Date()))
            build()
        }
        val line = obj.toString()
        writeExecutor.execute {
            try {
                val current = currentLogFile() ?: return@execute
                if (current.exists() && current.length() > ROTATION_THRESHOLD_BYTES) {
                    val prev = previousLogFile()
                    prev?.delete()
                    if (prev != null) current.renameTo(prev)
                }
                current.appendText(line + "\n", Charsets.UTF_8)
            } catch (e: Exception) {
                Log.w(TAG, "event log write failed", e)
            }
        }
    }
}
